async-41: Better Octree Performance

pwinston opened the pull request:
# Description
* Add canceling in-progress loads if no longer in view.
    * We had this for single-scale but never octree. 
* Add priority scheme to load 3, 2 and 1 level up first.
* Fix "double poll" issue in `QtPoll` with new `IntervalTimer`.
* Add event so visual can "wake up" QtPoll when new data is loaded.
    * Because new data might be loaded "out of the blue"
    * With no camera movement or loading happening at all.
    * But we need to start polling to load the chunks into VRAM.

# Demo 1

`NAPARI_OCTREE=1 napari https://s3.embassy.ebi.ac.uk/idr/zarr/v0.1/4495402.zarr`

Which requires `pip install ome-zarr`.

![napari-octree](https://user-images.githubusercontent.com/4163446/102445538-b02bd000-3ff9-11eb-8970-055745e97e18.gif)

This dataset takes like 1-2 seconds to load a single tile. So if you move enough, you have to wait that long since all workers will be loading stale data. We can cancel queued loads, but not ones already in progress, at least today.  It's impossible to cancel arbitrary workers, but we might be able to cancel network requests but it won't be easy.

There are still some bugs where tiles don't load, but the  "cancel in-progress loads" and "priority loading" are at least roughly working now.

# Demo 2

`NAPARI_OCTREE=1 napari http://hyper-resolution.org/dzi/Rijksmuseum/SK-C-5/SK-C-5_VIS_20-um_2019-12-21.dzi`

Which requires `pip install napari-dzi-zarr`.

This dataset loads in like 250ms per tile.

![napari-octree-3](https://user-images.githubusercontent.com/4163446/102446990-0d755080-3ffd-11eb-8698-a0320b6a8576.gif)

# Demo 3

![napari-octree-2](https://user-images.githubusercontent.com/4163446/102446447-cd619e00-3ffb-11eb-81a3-62d853800e8b.gif)

# Webmon

* Much better frame rate plot in https://github.com/pwinston/webmon/pull/21

![frame-rate-2](https://user-images.githubusercontent.com/4163446/102003003-16e67c00-3cd0-11eb-9c3f-ef7f097dfaf5.gif)

# Type of change
- [x] New feature in experimental code

# How has this been tested?
- [x] Manual testing

pwinston committed:
Send time with messages. For better graphing.

pwinston committed:
Add IntervalTimer to fix double poll

pwinston committed:
Rename to load_chunk

pwinston committed:
Unlog

pwinston committed:
1 worker and no "auto sync"

This performs 60Hz for me on the astronaut zarr

pwinston committed:
Fix log statement so it doesn't crash

pwinston committed:
If not Python39 this won't be true

pwinston committed:
Handle Python pre-3.9 more gracefully

pwinston committed:
Comment

pwinston committed:
Log not print

pwinston committed:
Debug log; log location

pwinston committed:
Debug logging; return future from load

pwinston committed:
Add location and elapsed_ms

elapsed_ms is so we can time chunks "end to end" instead of just timing the load portion. This is not fully setup yet, but we've been
logging it.

This led us to discover that in-progress
loads no longer need by the octree
were not being cancelled. But we've
add support for that now, still debugging.

pwinston committed:
Set workers to 10 for now, move logging

workers=1 is best when Dask is super slow
on big levels. But we need to experiment
with how to set that automatically.

pwinston committed:
Big comment/cleanup pass and debugging

Stil debugging this but wanted to get
the cleanup and doc strings in.

pwinston committed:
Logging

Probably will remove the prefixes though if
 the napari.async.octree.slice log path
 is shown.

pwinston committed:
New "location"

Octree location (level ,row col) is all we need
to log for octree. Will revise pre-octree image
later

pwinston committed:
Log elapsed, still debugging this

pwinston committed:
Change print -> log

pwinston committed:
Log, comments

pwinston committed:
Fix since load_chunks now returns tuple

pwinston committed:
Only add permanent chunk once up front!

pwinston committed:
Comments, slight cleanup

pwinston committed:
Comment

pwinston committed:
Better logging

pwinston committed:
Move method down

pwinston committed:
Add on_chunk_loaded to loader

pwinston committed:
Better logging; remove "label" param -> cleaner

pwinston committed:
Better logging

pwinston committed:
Better logging

pwinston committed:
Fix logging to a file, to log the root

pwinston committed:
Better log message (timing)

pwinston committed:
Log message (timing)

pwinston committed:
Logging + comments

pwinston committed:
Split napari.async and napari.octree logging

pwinston committed:
Use debug log level to be consistent

pwinston committed:
Make logging more compact, clearer

pwinston committed:
Minor logging

pwinston committed:
Minor logging

pwinston committed:
Logging - to make sure this wasn't being called!

pwinston committed:
New get_ancestors for loader

pwinston committed:
Log more standard numpy shape (x, y)

pwinston committed:
Logging

pwinston committed:
Trying using __hash__ here

pwinston committed:
Big changes to handle priority ordering

pwinston committed:
Fix logging syntax

pwinston committed:
Add ANCESTOR_LEVELS, fix logging

pwinston committed:
Fix log msg

pwinston committed:
Custom exception since was confusing

pwinston committed:
Log futures and log cleanup

pwinston committed:
Log syntax

pwinston committed:
Log name napari.loader to match napari.octree

pwinston committed:
Initial ChunkSet and debugging futures

pwinston committed:
Need __eq__ method

pwinston committed:
Log "loading" flag

pwinston committed:
Shorter log message

pwinston committed:
Cleanup

pwinston committed:
Logging cleanup

pwinston committed:
Need __hash__

pwinston committed:
Add load_ms, much cleaner than the 'special' timer name

pwinston committed:
Log the load_ms

pwinston committed:
Doc strings and move ChunkSet to own file

pwinston committed:
Put logging in method then comment out for now

pwinston committed:
Factor out log_chunks; remove some logging

pwinston committed:
Visual needs to "wake up" QtPoll on loads...

pwinston committed:
Cleanup logging

pwinston committed:
Remove some log spam

pwinston committed:
Tweak comments

sofroniewn commented on napari/_vispy/experimental/vispy_tiled_image_layer.py:
```diff
@@ -71,13 +75,20 @@ class VispyTiledImageLayer(VispyImageLayer):
         # self.node which VispyBaseImage holds.
         super().__init__(layer, visual)
 
+        # Create events after the base class. We have a loaded event that
+        # QtPoll listens to. Because a chunk might be loaded when QtPoll is
+        # totally quiet, no mouse movement, no in-progress loading. We need
+        # to get the polling going so we can load the chunks over time.
+        self.events = EmitterGroup(source=self, auto_connect=True, loaded=None)
```
I don't love that we had to do this, but I see why it might be necessary. There had been talk of using more events to do communication from the vispy visual a while back, but we didn't end up needing it. We might go down that path again though, but can cross those bridge when we get to them!

pwinston replied:
The list of things that would go away if were just drawing constantly is growing. But I still like the idea that napari can go 100% quiet. But there might be some global way to achieve that. Here we are trying to be super precise so napari won't poll or draw even a single frame more than it has to.

Maybe there's a way to just draw all the time "if the window is active" and "the user is doing something". As long as it also keeps drawing until all loads have finished, even if the user is not doing anything. That's the main source of complexity. Prior to octree we only drew when the camera moved. Now loading is a type of "animation" that needs to finish even if the user has gone to get coffee.

But maybe we could have some global system. And anyone can just ping it say "keep going". Things do seem to be getting complicated just to workaround the issue that we are not always drawing.

sofroniewn replied:
> Things to do seem to be getting complicated just to workaround the issue that we are not always drawing.

Yeah, definitely something to consider, but agreed I do like that it can "go 100% quiet" too ....

sofroniewn reviewed (approved):
I'm approving. Changes in performance are very noticeable. I flagged the addition of the event emitter to the vispy visual which I think makes sense, but isn't something we needed before.

This PR is still in draft, so I'll wait to hear more from @pwinston before merging, but it looks good!

pwinston committed:
Don't poll right away, was leding to crashes

C++ object deleted

pwinston committed:
Fix log msg

pwinston committed:
Better log msg

pwinston committed:
Do create ancestors, so we can load them

pwinston committed:
Comments

pwinston committed:
Create self.rows and cols

pwinston committed:
Fix up chunk getting

pwinston committed:
Remove redundant logging

pwinston committed:
Comment

pwinston committed:
Better comment

pwinston committed:
No logging, comments

pwinston committed:
Comments

pwinston committed:
Re-order methods and cleanup

pwinston committed:
Shorten comment

pwinston committed:
Rename param; add logging

pwinston committed:
Only check size if creating

Because get_children can ask for
out of bounds chunks. Because it doesn't know about edges and corners
with non-power-of-two images

pwinston committed:
Move comment

pwinston committed:
Fix test: needed future/tuple

pwinston committed:
Fix test: no future for that call

pwinston committed:
Remove log spam

pwinston committed:
Add expansion factor
